functions/src/business-logics/user-logics.ts

```diff
@@ -2,3 +2,48 @@
 //  allow users to update their firstName, lastName, birthDate, and avatarUrl
 //  avatarUrl must be saved to users/{uid}/profile-pictures path in storage
 //  create a security fn that will prevent users from changing their username
+
+import {Entity} from "../db-structure";
+import {LogicConfig, LogicResultDoc} from "emberflow/lib/types";
+import {getFileDataFromUrl, moveStorageFile} from "./utils";
+
+export const onUserUpdateLogic: LogicConfig = {
+  name: "onUserUpdateLogic",
+  actionTypes: ["update"],
+  modifiedFields: ["firstName", "lastName", "birthDate", "avatarUrl"],
+  entities: [Entity.User],
+  logicFn: async (action) => {
+    const {
+      eventContext: {
+        docPath,
+        docId,
+      },
+      modifiedFields,
+    } = action;
+
+    const {avatarUrl} = modifiedFields;
+
+    if (avatarUrl) {
+      const {fileExtension} = getFileDataFromUrl(avatarUrl);
+      const fileFullname = `${docId}.${fileExtension}`;
+      const fileDstPath = `users/${docId}/profile-pictures/${fileFullname}`;
+      modifiedFields.avatarUrl = fileDstPath;
+      await moveStorageFile(avatarUrl, fileDstPath);
```

Comment:
Actually I realized there's something missing here. Supposedly, you should also delete the previous avatar url by using the deleteStorageFile().

you can browse the onPostCreateLogic for sample

Here's how it works: First, check if there is an existing avatar image in the document. Then, compare the file types (you can use the file extensions) of the old and new avatar URLs. Only run deleteStorageFile() if the file types are different.

The reason for this is that moveStorageFile will automatically overwrite the old avatar URL if the new avatar has the same file name and type. Since the docId is used as the file name, we can assume it will be identical for both the old and new avatars.